From b4d49d344fc981c5dee1fb743cf0dc25b94b0719 Mon Sep 17 00:00:00 2001
From: "Hebbar, Gururaja" <gururaja.hebbar@ti.com>
Date: Wed, 11 May 2016 10:17:34 -0500
Subject: [PATCH 182/347] Input: matrix-keypad: Enhance pinctrl support

This patch comes from:
  git://git.ti.com/processor-sdk/processor-sdk-linux.git

Update matrix-keypad driver to set the state of the pins to:
- "sleep" on suspend
- "default" on resume

By optionally putting the pins into sleep state in the suspend callback
we can accomplish two things.
- minimize current leakage from pins and thus save power,
- prevent the IP from driving pins output in an uncontrolled manner,
which may happen if the power domain drops the domain regulator.

When using any of the GPIOs in a matrix-keypad for wakeup the pins must
be set to the appropriate state to allow this dependent on the board.

Signed-off-by: Hebbar, Gururaja <gururaja.hebbar@ti.com>
Signed-off-by: Dave Gerlach <d-gerlach@ti.com>
Signed-off-by: Sekhar Nori <nsekhar@ti.com>
(cherry picked from commit 3e4a63f190461a1478477835daf77079eb7133be)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/input/keyboard/matrix_keypad.c |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/drivers/input/keyboard/matrix_keypad.c b/drivers/input/keyboard/matrix_keypad.c
index 7f12b65..fe02f9f 100644
--- a/drivers/input/keyboard/matrix_keypad.c
+++ b/drivers/input/keyboard/matrix_keypad.c
@@ -25,6 +25,7 @@
 #include <linux/of.h>
 #include <linux/of_gpio.h>
 #include <linux/of_platform.h>
+#include <linux/pinctrl/consumer.h>
 
 struct matrix_keypad {
 	const struct matrix_keypad_platform_data *pdata;
@@ -280,6 +281,8 @@ static int matrix_keypad_suspend(struct device *dev)
 	if (device_may_wakeup(&pdev->dev))
 		matrix_keypad_enable_wakeup(keypad);
 
+	pinctrl_pm_select_sleep_state(dev);
+
 	return 0;
 }
 
@@ -291,6 +294,8 @@ static int matrix_keypad_resume(struct device *dev)
 	if (device_may_wakeup(&pdev->dev))
 		matrix_keypad_disable_wakeup(keypad);
 
+	pinctrl_pm_select_default_state(dev);
+
 	matrix_keypad_start(keypad->input_dev);
 
 	return 0;
-- 
1.7.5.4

